arrow-libs/core/arrow-core-retrofit/src/main/kotlin/arrow/retrofit/adapter/either/networkhandling/NetworkEitherCallAdapter.kt

```diff
@@ -0,0 +1,91 @@
+package arrow.retrofit.adapter.either.networkhandling
+
+import arrow.core.Either
+import arrow.core.left
+import arrow.core.right
+import okhttp3.Request
+import okio.Timeout
+import retrofit2.Call
+import retrofit2.CallAdapter
+import retrofit2.Callback
+import retrofit2.Response
+import java.io.IOException
+import java.lang.reflect.Type
+
+internal class NetworkEitherCallAdapter<R>(
+  private val successType: Type
+) : CallAdapter<R, Call<Either<CallError, R?>>> {
+
+  override fun adapt(call: Call<R?>): Call<Either<CallError, R?>> = EitherCall(call, successType)
+
+  override fun responseType(): Type = successType
+}
+
+private class EitherCall<R>(
+  private val delegate: Call<R>,
+  private val successType: Type
+) : Call<Either<CallError, R>> {
+
+  override fun enqueue(callback: Callback<Either<CallError, R>>) = delegate.enqueue(
+    object : Callback<R> {
+
+      override fun onResponse(call: Call<R>, response: Response<R>) {
+        callback.onResponse(this@EitherCall, Response.success(response.toEither()))
+      }
+
+      private fun Response<R>.toEither(): Either<CallError, R> {
+        // Http error response (4xx - 5xx)
+        if (!isSuccessful) {
+          val errorBody = errorBody()!!.string()
```

Comment:
IMO
```kotlin
val errorBody = checkNotNull(errorBody()) { "Response.errorBody() must be not null!" }.string()
```

Reply:
Yes, I was also thinking about this. But it's impossible for `errorBody` to be null, unless a bug is introduced in Retrofit (in which case not only this adapter will have problems, but the whole Retrofit based world will break in some way).

`Response<T>` has fields `@Nullable T body` and `@Nullable ResponseBody errorBody`. The reason they are both nullable is that always only one of them is not null. In case of an error response (which in my adapter is checked by `if (!isSuccessful)`, `errorBody` is guaranteed by Retrofit implementation logic to be non-null. Look at the [implementation](https://github.com/square/retrofit/blob/9b32c965ef31275627e35a862f584bcc7b9fa4e5/retrofit/src/main/java/retrofit2/Response.java#L90) of both `Response.error()` methods - they both already perform the check.

Reply:
That makes sense